src/main/java/org/mule/extensions/jms/internal/source/JmsMessageListener.java

```diff
@@ -139,17 +132,41 @@ private void dispatchMessage(Message message, SourceCallbackContext context, Str
     try {
       Result<Object, JmsAttributes> result = resultFactory.createResult(message, jmsSupport.getSpecification(),
                                                                         contentType, encoding, session.getAckId());
-      evaluateAckAction(message);
+
+      registerAckIdIfRequired(message);
+
       sourceCallback.handle(result, context);
 
+      doAckIfRequired(message);
```

Comment:
Add test coverage for ack, nack scenarios depending on if sourceCallback fails or not.